Complete postproccessdkg

wrinkledeth opened the pull request:
Completed postProcessDkg(), slashNode(), freezeNode(), nodeStake(), and nodeUnstake() + relevant tests.

kafeikui commented on contracts/src/Controller.sol:
```diff
@@ -707,12 +687,13 @@ contract Controller is Ownable {
-    // 1. emit GroupRelayTask if grouping successfully
-    // 2. arrange members if fail to group
-    // and rewards trigger (sender)
+    event groupRelayTask(
```
We don't involve relay feature in the first version, so this can be removed.

wrinkledeth replied:
done

kafeikui commented on contracts/src/Controller.sol:
```diff
@@ -740,41 +721,255 @@ contract Controller is Ownable {
+                groupIndex
+            );
+
+        if (!isStrictlyMajorityConsensusReached) {
```
I can't see the reason to reverse the if condition, we need to keep the former logic. Since we don't have to do anything like emit relay event on isStrictlyMajorityConsensusReached, then the code should be like
```
if (!isStrictlyMajorityConsensusReached) {
    ...do arrange member things...
}
rewards[msg.sender] += DKG_POST_PROCESS_REWARD;
```

wrinkledeth replied:
you're right, the condition should not have been reversed. This has been completed.

kafeikui commented on contracts/src/Controller.sol:
```diff
@@ -740,41 +721,255 @@ contract Controller is Ownable {
+                uint256 minimum = minimumThreshold(g.size);
+
+                // set g.threshold to max (default min threshold / minimum threshold)
+                g.threshold = g.threshold > minimum
```
not g.threshold
```
g.threshold = DEFAULT_MINIMUM_THRESHOLD > minimum
                    ? DEFAULT_MINIMUM_THRESHOLD
                    : minimum;
```

wrinkledeth replied:
done

kafeikui commented on contracts/src/Controller.sol:
```diff
@@ -740,41 +721,255 @@ contract Controller is Ownable {
+                        DISQUALIFIED_NODE_PENALTY_AMOUNT,
+                        0, //! should this be 0?
+                        handleGroup
+                    );
```
We need to [arrange group members when post_process_dkg even if disqualified_nodes is empty](https://github.com/ARPA-Network/BLS-TSS-Network/commit/e8b475b7f503020c5a85cd650440127cffa8d0a7), please refer to #14

wrinkledeth replied:
Okay I see, so I should implement arrange_members_in_group() in solidity and include it at the end of post_proccess_dkg()?

wrinkledeth replied:
This has been completed!

kafeikui commented on contracts/src/Controller.sol:
```diff
@@ -740,41 +721,255 @@ contract Controller is Ownable {
+                            break;
+                        }
+                    }
+                }
```
We need to handle the possible gaps in g.members by `delete g.members[i]` during or after the for loop. Element order in the array is not that important, so we can swap with the last and pop the last, like the for loop you wrote in line 499-509. Also, make sure the outer loop iterating `disqualifiedNodes` as it won't affect `delete` element in `g.members` during loop, and possibly save cycles.

wrinkledeth replied:
To confirm this what you are looking for right?

```js
                for (uint256 j = 0; j < disqualifiedNodes.length; j++) {
                    for (uint256 i = 0; i < g.members.length; i++) {
                        if (g.members[i].nodeIdAddress == disqualifiedNodes[j]) {
                            g.members[i] = g.members[g.members.length - 1];
                            g.members.pop();
                            break;
                        }
                    }
                }
```

wrinkledeth replied:
completed

kafeikui commented on contracts/src/Controller.sol:
```diff
@@ -740,41 +721,255 @@ contract Controller is Ownable {
+                node.staking - unstakeAmount >= NODE_STAKING_AMOUNT,
+                "Node state is true, cannot unstake below staking threshold"
+            );
+        }
```
That's ok, and we also need an external `node_quit` function.

kafeikui commented on contracts/src/Controller.sol:
```diff
@@ -740,41 +721,255 @@ contract Controller is Ownable {
+            );
+        }
+
+        // require(node.staking - unstakeAmount >= 0, "Cannot unstake more than amount staked"); //need this?
```
Not really, evm will revert if there is over/underflow on uint over solidity 0.8.

wrinkledeth replied:
understood.

kafeikui commented on contracts/src/Controller.sol:
```diff
@@ -740,41 +721,255 @@ contract Controller is Ownable {
+                        }
+                    }
+
+                    // try to reblance each group, if a any fail, set rebalanceFailure to true
```
The behavior should be described like "try to reblance each group, if succeed in rebalancing, break the loop and set rebalanceFailure to false". Almost like code at line108-112. Also, groupIndices[i] is right instead of i.
```
            bool rebalanceFailure = true;
            for (uint256 i = 0; i < groupIndices.length; i++) {
                if (rebalanceGroup(groupIndices[i], groupIndex)) {
                    rebalanceFailure = false;
                    break;
                }
            }
```

wrinkledeth replied:
Makes sense, thank you!

kafeikui commented on contracts/src/Controller.sol:
```diff
@@ -740,41 +721,255 @@ contract Controller is Ownable {
+                                bool _needsRebalance
+                            ) = findOrCreateTargetGroup();
+
+                            delete (_needsRebalance); //! Used to supress "unused parameter" warning, should I just leave this out?
```
We can destructure return value and ignore some value in solidity too. 

(uint256 targetGroupIndex, ) = findOrCreateTargetGroup();

wrinkledeth replied:
Wow! Thank you for this.

kafeikui commented on contracts/src/Controller.sol:
```diff
@@ -740,41 +721,255 @@ contract Controller is Ownable {
+                                false
+                            );
+
+                            // if group at targetGroupIndex now has 3+ members, emit group event
```
I think it's better to follow the rust mock version, as `emitGroupEvent` will prepare all the environment on that group as well as building coordinator and `emit dkgTask`, we don't want that happen on the same group repeatedly but only on the latest state.

wrinkledeth replied:
Apologies, I see my mistake now. I have corrected it to match the rust mock contract.

kafeikui reviewed (changes requested):
Besides comments on Controller.sol(src), we need to cover test case on `commitDkg/postProcessDkg` with different `CommitResult`(groupEpoch, publicKey and disqualifiedNodes), e.g. all identical, happy path(as we already have), there is no strict majority CommitResult, there is a strictly majority CommitResult but not exceeding threshold and finally, there is a strictly majority CommitResult and exceeding threshold but not all identical.

wrinkledeth commented on contracts/src/Controller.sol:
```diff
@@ -740,41 +721,255 @@ contract Controller is Ownable {
+                if (needsRebalance) {
+                    // set isStrictlyMajorityConsensusReached to false
+                    groups[groupIndex]
+                        .isStrictlyMajorityConsensusReached = false;
```
To clarify, do you mean that:

```
                    groups[groupIndex]
                        .isStrictlyMajorityConsensusReached = false;
```
                        
Should be moved to 907?

kafeikui replied:
yes, then we will set `isStrictlyMajorityConsensusReached` of the current group to false no matter succeed in rebalancing.

wrinkledeth replied:
done

wrinkledeth commented on contracts/src/Controller.sol:
```diff
@@ -740,41 +721,255 @@ contract Controller is Ownable {
+                            rebalanceFailure = true;
+                        }
+                    }
+                    // if rebalance failed, start DKG for groupIndex
```
I think i noticed a potential bug here. 

When groupCount == 1:

groupIndices will be length 0, and so lines 901-903 will not run, and rebalanceFailure can never be set to true.

What should happen when there is only 1 group with 3 members. When we call freezeNode on a node, that node will get removed group the group, and we fall below 3 members ( the default minimum threshold). Which means we should set isStrictylMajorityConsensus reached to False, right?

Wondering if we should write an edgecase to handle when groupCount == 1?

wrinkledeth replied:
This has been resolved, please ignore the above.

wrinkledeth commented on contracts/src/Controller.sol:
```diff
@@ -740,41 +721,255 @@ contract Controller is Ownable {
+
+                            // if the current group index is selected, break
+                            if (groupIndex == targetGroupIndex) {
+                                break;
```
Should this be continue instead?

wrinkledeth replied:
sorry, ignore this please.

wrinkledeth commented:
Latest commit addresses the issues raised above and includes the following changes: 
- bug fixes in postProccessDkg() and freezeNode()
- arrangeMembersInGroup() implemented. 
- nodeQuit() implemented (still need erc20 interaction). 
- Fixed bug in removeFromGroup() when group.size == 0
-  cleaned up comments / changed formatting to Forge instead of Prettier.

Upcoming work planned for next PR: 
- add erc20 interaction to nodeStake and nodeQuit,
- consolidate threshold logic to emitGroupEvent,
- writing proper tests for postproccessdkg / freezenode.

kafeikui commented on contracts/src/Controller.sol:
```diff
@@ -725,36 +620,253 @@ contract Controller is Ownable {
+            uint256[] memory involvedGroupsFinal = new uint256[](numInvolvedGroups);
+            for (uint256 i = 0; i < numInvolvedGroups; i++) {
+                involvedGroupsFinal[i] = involvedGroups[i];
+            }
```
We can use a bitmap to avoid removing duplicate. Since we are dealing it in memory, a bool[] should be ok.

kafeikui replied:
I think we can merge the pr first then improve it.
